File: relay/throttle.go
package main

import (
	"strings"
	"sync"
	"time"
)

// authThrottle is the application-level online brute-force throttle for the
// submission auth door (#105, audit F2). It is keyed on the presented ACCOUNT, not
// the source IP: behind the bastion every public connection presents one
// source IP (the bastion masquerade), so per-IP throttling is blind. A second,
// GLOBAL layer bounds spread-account spraying (one guess each across many accounts)
// through that single IP.
//
// Properties:
//   - fail-CLOSED + enumeration-SAFE: a throttled attempt returns the SAME generic
//     auth failure as a wrong password, and existent + non-existent usernames are
//     throttled identically, so the throttle never reveals whether an account
//     exists (it only ever reveals "you have been failing on this name", which the
//     attacker already knows).
//   - only real password rejections count; an infra error (backend down) does NOT,
//     so an outage cannot lock users out.
//   - in-memory + per-process. The submission daemon is a single process, so this
//     is sufficient; a multi-instance deploy would need a shared store (documented).
//
// All methods are safe on a nil receiver and no-op when disabled, so a session
// built without a throttle (tests) behaves as before.
type authThrottle struct {
	enabled      bool
	maxFailures  int           // per-account consecutive failures before lockout
	lockout      time.Duration // base per-account lockout; doubles per failure past the threshold
	maxLockout   time.Duration // cap on the per-account backoff window (also the idle-decay window)
	globalMax    int           // aggregate failures within globalWindow before a global cooldown (0 = off)
	globalWindow time.Duration

	now func() time.Time // injectable clock (tests)

	mu          sync.Mutex
	accounts    map[string]*acctFailState
	globalCount int
	globalStart time.Time
	globalUntil time.Time
}

type acctFailState struct {
	failures    int
	lastFailure time.Time
	lockedUntil time.Time
}

// newAuthThrottle builds a throttle from config. A nil clock defaults to time.Now.
// Knobs are clamped to sane minimums so a misconfig cannot disable the control
// while leaving it "enabled".
func newAuthThrottle(cfg ThrottleCfg, now func() time.Time) *authThrottle {
	if now == nil {
		now = time.Now
	}
	maxFailures := cfg.MaxFailures
	if maxFailures < 1 {
		maxFailures = 5
	}
	lockout := cfg.Lockout
	if lockout <= 0 {
		lockout = 60 * time.Second
	}
	maxLockout := cfg.MaxLockout
	if maxLockout < lockout {
		maxLockout = lockout
	}
	globalWindow := cfg.GlobalWindow
	if globalWindow <= 0 {
		globalWindow = 60 * time.Second
	}
	return &authThrottle{
		enabled:      cfg.Enabled,
		maxFailures:  maxFailures,
		lockout:      lockout,
		maxLockout:   maxLockout,
		globalMax:    cfg.GlobalMax,
		globalWindow: globalWindow,
		now:          now,
		accounts:     make(map[string]*acctFailState),
	}
}

// throttleKey normalizes a username into the throttle's account key: lower-cased
// and trimmed, so trivial case variations cannot multiply the failure budget.
func throttleKey(username string) string {
	return strings.ToLower(strings.TrimSpace(username))
}

// allow reports whether an auth attempt for account may proceed to the backend. It
// returns false when a global cooldown is in effect OR the account is in a
// per-account lockout. A denial is treated by the caller as a generic auth failure.
func (t *authThrottle) allow(account string) bool {
	if t == nil || !t.enabled {
		return true
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	now := t.now()
	if now.Before(t.globalUntil) {
		return false
	}
	st := t.accounts[account]
	if st == nil {
		return true
	}
	return !now.Before(st.lockedUntil)
}

// fail records a real password rejection for account and updates the per-account
// and global lockouts.
func (t *authThrottle) fail(account string) {
	if t == nil || !t.enabled {
		return
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	now := t.now()

	st := t.accounts[account]
	if st == nil {
		st = &acctFailState{}
		t.accounts[account] = st
	}
	// Idle decay: a long-quiet account starts fresh (bounds memory and avoids
	// escalating against failures spread hours apart).
	if !st.lastFailure.IsZero() && now.Sub(st.lastFailure) > t.maxLockout {
		st.failures = 0
		st.lockedUntil = time.Time{}
	}
	st.failures++
	st.lastFailure = now
	if st.failures >= t.maxFailures {
		st.lockedUntil = now.Add(t.backoff(st.failures))
	}

	// Global layer: count failures within a sliding window; once the ceiling is
	// crossed, cool down ALL auth for one window (spread-spraying backstop).
	if now.Sub(t.globalStart) > t.globalWindow {
		t.globalCount = 0
		t.globalStart = now
	}
	t.globalCount++
	if t.globalMax > 0 && t.globalCount > t.globalMax {
		t.globalUntil = now.Add(t.globalWindow)
	}

	t.pruneLocked(now)
}

// success clears an account's failure state (a correct password fully resets it).
func (t *authThrottle) success(account string) {
	if t == nil || !t.enabled {
		return
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	delete(t.accounts, account)
}

// backoff is the per-account lockout duration: the base doubled once per failure
// beyond the threshold, capped at maxLockout.
func (t *authThrottle) backoff(failures int) time.Duration {
	d := t.lockout
	for i := 0; i < failures-t.maxFailures && d < t.maxLockout; i++ {
		d *= 2
	}
	if d > t.maxLockout {
		d = t.maxLockout
	}
	return d
}

// pruneLocked drops idle account entries so the map cannot grow unbounded under a
// username-spraying attack. Called under lock; only scans once the map is large.
func (t *authThrottle) pruneLocked(now time.Time) {
	if len(t.accounts) < 1024 {
		return
	}
	for k, st := range t.accounts {
		if now.Sub(st.lastFailure) > t.maxLockout && now.After(st.lockedUntil) {
			delete(t.accounts, k)
		}
	}
}

File: imap/posternimap/throttle.py
"""Application-level online brute-force throttle for the IMAP auth door (#105).

The Python mirror of the SMTP relay's throttle (relay/throttle.go), built to the
RATIFIED cross-door contract (#105): identical AUTH_THROTTLE_* knobs (integer
seconds) and the two behavioural invariants both doors enforce:

  1. Only a real password rejection counts. An infra/backend error (store or
     directory down) must NOT count, so an outage can never lock users out. This
     module only advances state when the caller invokes fail(); the caller calls
     fail() solely on a genuine bad-credential result, never on a backend fault.
  2. A throttled/locked attempt must be INDISTINGUISHABLE from a normal auth
     failure (no account-existence leak). allow() only returns False; the caller
     returns the same generic UnauthorizedLogin it returns for a wrong password,
     and existent + non-existent usernames are throttled identically.

Keyed on the presented ACCOUNT, not the source IP: behind the bastion
every external connection presents one masqueraded source IP, so per-IP is blind.
A second GLOBAL layer bounds spread-account spraying through that single IP.

In-memory + per-process (one portal per proxy process), matching the Go side; a
multi-instance deploy would need a shared store (documented). Disabled = no-op.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass
from typing import Callable, Dict, Optional


def throttle_key(username: str) -> str:
    """Normalize a username into the throttle key: trimmed + lower-cased, so case
    variations cannot multiply the failure budget (mirrors relay throttleKey)."""
    return username.strip().lower()


@dataclass
class _AcctState:
    failures: int = 0
    last_failure: Optional[float] = None  # None == never failed
    locked_until: float = 0.0


class AuthThrottle:
    """Per-account + global auth throttle. All methods are safe when disabled
    (no-op) so a portal built without throttling behaves as before."""

    def __init__(
        self,
        *,
        enabled: bool,
        max_failures: int,
        lockout_seconds: float,
        max_lockout_seconds: float,
        global_max_failures: int,
        global_window_seconds: float,
        now: Optional[Callable[[], float]] = None,
    ) -> None:
        # Clamp to sane minimums so a misconfig cannot leave the control "enabled"
        # but toothless (mirrors newAuthThrottle).
        self._enabled = enabled
        self._max_failures = max_failures if max_failures >= 1 else 5
        self._lockout = lockout_seconds if lockout_seconds > 0 else 60.0
        self._max_lockout = max_lockout_seconds if max_lockout_seconds >= self._lockout else self._lockout
        self._global_max = global_max_failures
        self._global_window = global_window_seconds if global_window_seconds > 0 else 60.0
        # Monotonic by default: immune to wall-clock jumps (NTP / DST).
        self._now = now or time.monotonic
        self._lock = threading.Lock()
        self._accounts: Dict[str, _AcctState] = {}
        self._global_count = 0
        self._global_start = 0.0
        self._global_until = 0.0

    def allow(self, account: str) -> bool:
        """True if an attempt for account may proceed to the backend. False under a
        global cooldown OR a per-account lockout; the caller treats False as a
        generic auth failure (invariant 2)."""
        if not self._enabled:
            return True
        with self._lock:
            now = self._now()
            if now < self._global_until:
                return False
            st = self._accounts.get(account)
            if st is None:
                return True
            return not (now < st.locked_until)

    def fail(self, account: str) -> None:
        """Record a real password rejection (invariant 1: callers never invoke this
        for an infra/backend fault) and update per-account + global lockouts."""
        if not self._enabled:
            return
        with self._lock:
            now = self._now()
            st = self._accounts.get(account)
            if st is None:
                st = _AcctState()
                self._accounts[account] = st
            # Idle decay: a long-quiet account starts fresh (bounds memory, avoids
            # escalating across failures spread hours apart).
            if st.last_failure is not None and (now - st.last_failure) > self._max_lockout:
                st.failures = 0
                st.locked_until = 0.0
            st.failures += 1
            st.last_failure = now
            if st.failures >= self._max_failures:
                st.locked_until = now + self._backoff(st.failures)
            # Global layer: failures within a sliding window; crossing the ceiling
            # cools down ALL auth for one window (spread-spraying backstop).
            if (now - self._global_start) > self._global_window:
                self._global_count = 0
                self._global_start = now
            self._global_count += 1
            if self._global_max > 0 and self._global_count > self._global_max:
                self._global_until = now + self._global_window
            self._prune_locked(now)

    def success(self, account: str) -> None:
        """A correct password fully resets the account's failure state."""
        if not self._enabled:
            return
        with self._lock:
            self._accounts.pop(account, None)

    def _backoff(self, failures: int) -> float:
        # base, doubled once per failure beyond the threshold, capped at maxLockout.
        d = self._lockout
        i = 0
        while i < failures - self._max_failures and d < self._max_lockout:
            d *= 2
            i += 1
        return min(d, self._max_lockout)

    def _prune_locked(self, now: float) -> None:
        # Drop idle entries so the map cannot grow unbounded under username spraying.
        if len(self._accounts) < 1024:
            return
        for k in list(self._accounts):
            st = self._accounts[k]
            if st.last_failure is not None and (now - st.last_failure) > self._max_lockout and now > st.locked_until:
                del self._accounts[k]


def build_throttle(cfg, now: Optional[Callable[[], float]] = None) -> AuthThrottle:
    """Construct the throttle from Config (AUTH_THROTTLE_* knobs)."""
    return AuthThrottle(
        enabled=cfg.throttle_enabled,
        max_failures=cfg.throttle_max_failures,
        lockout_seconds=cfg.throttle_lockout_seconds,
        max_lockout_seconds=cfg.throttle_max_lockout_seconds,
        global_max_failures=cfg.throttle_global_max_failures,
        global_window_seconds=cfg.throttle_global_window_seconds,
        now=now,
    )
